crates/term/src/widget/mod.rs

```diff
@@ -0,0 +1,531 @@
+        };
+
+        if blink {
+            let sin_component = self.stable_time / std::f32::consts::FRAC_PI_2 * 13.;
+            let alpha = (sin_component.sin() + 1.) / 2.;
+            inner_color = inner_color.gamma_multiply(alpha);
+        }
+
+        let cursor = galley.from_rcursor(egui::epaint::text::cursor::RCursor {
+            row: cursor_point.line.0 as usize,
+            column: cursor_point.column.0,
+        });
+
+        let mut cursor_pos = galley.pos_from_cursor(&cursor).min + response.rect.min.to_vec2();
+
+        let cursor_rect = match cursor_shape {
+            CursorShape::Block | CursorShape::HollowBlock | CursorShape::Hidden => {
+                egui::Rect::from_min_size(cursor_pos, egui::vec2(char_width, row_height))
+            }
+            CursorShape::Beam => egui::Rect::from_min_size(cursor_pos, egui::vec2(2.0, row_height)),
+            CursorShape::Underline => {
+                cursor_pos.y += row_height - 2.0;
+                egui::Rect::from_min_size(cursor_pos, egui::vec2(char_width, 2.0))
+            }
+        };
+
+        painter.rect(
+            cursor_rect,
+            egui::Rounding::ZERO,
+            inner_color,
+            egui::Stroke::new(1.0, outer_color),
+        );
+
+        // scrollbar
+        let min = response.rect.min + egui::vec2(galley.rect.width(), 0.0);
+        let size = egui::vec2(5., response.rect.height());
+        let sidebar_rect = egui::Rect::from_min_size(min, size);
+        painter.rect_filled(
+            sidebar_rect,
+            egui::Rounding::ZERO,
+            ui.visuals().extreme_bg_color,
+        );
+
+        // handle
+        let scrollbar_percent = screen_lines as f32 / total_lines as f32;
+        let scrollbar_height = scrollbar_percent * response.rect.height();
+
+        let scrollbar_offset_percent = display_offset as f32 / total_lines as f32;
+        let scrollbar_offset = scrollbar_offset_percent * response.rect.height();
+        let scrollbar_y = response.rect.max.y - scrollbar_offset - scrollbar_height;
+
+        let min = egui::pos2(min.x, scrollbar_y);
+        let size = egui::vec2(5., scrollbar_height);
+        let scrollbar_rect = egui::Rect::from_min_size(min, size);
+
+        painter.rect_filled(
+            scrollbar_rect,
+            egui::Rounding::same(5.),
+            ui.visuals().widgets.active.fg_stroke.color,
+        );
+
+        painter
+            .ctx()
+            .request_repaint_after(std::time::Duration::from_millis(16));
+
+        if response.has_focus() {
+            ui.memory_mut(|mem| mem.set_focus_lock_filter(response.id, FILTER));
+
+            let (events, modifiers) = ui.input(|i| (i.filtered_events(&FILTER), i.modifiers));
+            self.process_egui_events(
+                events,
+                modifiers,
+                response.rect.min,
+                response.hover_pos(),
+                &galley,
+            );
+        }
```

Comment:
Please also set the IME info when the terminal is focused so that input method editors will start composition when typing text into the terminal:

```rust
if response.has_focus() {
    ui.output_mut(|o| {
        o.ime = Some(egui::output::IMEOutput {
            // ...
        });
    });
}
```
The `egui::output::IMEOutput` has two fields: `rect` which should be the `egui::Rect` of the editable part of the terminal and `cursor_rect` which should be the `egui::Rect` of the cursor in the terminal ***ignoring scroll*** (since the terminal scrolls to the bottom when typing).

Reply:
I think I may have fixed this, but I'm not sure how to set up an IME on my system with a graphical display. I did test it with IBus + SKK and it seemed to work, though

Reply:
It does open the IME window now but the text doesn't get typed into the terminal. I'll fix this later though since you don't have a working IME.

Reply:
I think it might be because it's not handling `egui::Event::CompositionEnd` right now?

Reply:
Ok, I added some basic handling of IME.
![image](https://github.com/Astrabit-ST/Luminol/assets/54482069/c7ec9ced-04ed-4ead-a052-1e751b8fba01)
I'm not super happy with how it works (displaying at the top right, rather than at the cursor) but it'll do for now